File: leetcode/1-100/9-palindrome-number/bench/palindrome.py
"""Benchmark workload — half-reverse Palindrome Number. Python mirror
of bench/palindrome.kara. Same N, K, same LCG fill + every-16th
manufactured 8-digit palindrome, same sink formula — see that file's
header for the workload rationale.

Python's int is arbitrary-precision, so inputs are normalized into the
signed 32-bit range with a sign-bit mask via `to_i32`. Unlike kata #7,
the early `x < 0` reject means the palindrome loop only ever sees
non-negative `x`, so Python's floor `//` and Kāra's truncated `/`
produce the same results — no c_div / c_mod helpers needed.
"""

from __future__ import annotations

INT_MAX = 2**31 - 1
INT_MIN = -(2**31)


def is_palindrome(x: int) -> bool:
    if x < 0 or (x % 10 == 0 and x != 0):
        return False
    reversed_ = 0
    while x > reversed_:
        reversed_ = reversed_ * 10 + x % 10
        x = x // 10
    return x == reversed_ or x == reversed_ // 10


def to_i32(v: int) -> int:
    v &= 0xFFFFFFFF
    return v - 0x100000000 if v & 0x80000000 else v


def manufacture_palindrome(v32: int) -> int:
    lo = -v32 if v32 < 0 else v32
    four_raw = lo % 10000
    four = four_raw + 1000 if four_raw < 1000 else four_raw
    d0 = four % 10
    d1 = (four // 10) % 10
    d2 = (four // 100) % 10
    d3 = (four // 1000) % 10
    rev = d0 * 1000 + d1 * 100 + d2 * 10 + d3
    return four * 10000 + rev


def main() -> None:
    n = 1024
    k_iters = 50_000_000

    inputs: list[int] = []
    for i in range(n):
        v32 = to_i32(i * 2_654_435_769 + 305_419_896)
        if i % 16 == 0:
            inputs.append(manufacture_palindrome(v32))
        else:
            inputs.append(v32)

    total = 0
    for k in range(k_iters):
        x = inputs[k % n]
        if is_palindrome(x):
            total += 1
    print(total)


if __name__ == "__main__":
    main()

File: leetcode/1-100/9-palindrome-number/bench/palindrome.rs
// Benchmark workload — half-reverse Palindrome Number. Rust mirror of
// bench/palindrome.kara. Same N, K, same LCG fill + every-16th
// manufactured 8-digit palindrome, same sink formula — see that file's
// header for the workload rationale.

fn is_palindrome(x: i32) -> bool {
    if x < 0 || (x % 10 == 0 && x != 0) {
        return false;
    }
    let mut x = x;
    let mut reversed: i32 = 0;
    while x > reversed {
        reversed = reversed * 10 + x % 10;
        x /= 10;
    }
    x == reversed || x == reversed / 10
}

fn manufacture_palindrome(v32: i32) -> i32 {
    let lo = if v32 < 0 { -v32 } else { v32 };
    let four_raw = lo % 10000;
    let four = if four_raw < 1000 { four_raw + 1000 } else { four_raw };
    let d0 = four % 10;
    let d1 = (four / 10) % 10;
    let d2 = (four / 100) % 10;
    let d3 = (four / 1000) % 10;
    let rev = d0 * 1000 + d1 * 100 + d2 * 10 + d3;
    four * 10000 + rev
}

fn main() {
    let n: i64 = 1024;
    let k_iters: i64 = 50_000_000;

    let mut inputs: Vec<i32> = Vec::with_capacity(n as usize);
    for i in 0..n {
        let raw: i64 = i.wrapping_mul(2_654_435_769).wrapping_add(305_419_896);
        let v32: i32 = raw as i32;
        if i % 16 == 0 {
            inputs.push(manufacture_palindrome(v32));
        } else {
            inputs.push(v32);
        }
    }

    let mut sum: i64 = 0;
    for k in 0..k_iters {
        let idx = (k % n) as usize;
        let x = inputs[idx];
        sum += if is_palindrome(x) { 1 } else { 0 };
    }
    println!("{}", sum);
}
